catalyst-toolbox-cli/src/cli/recovery/tally/mockchain.rs

```diff
@@ -0,0 +1,108 @@
+use chain_impl_mockchain::block::HeaderId;
+use chain_impl_mockchain::fragment::Fragment;
+use chain_impl_mockchain::ledger::Ledger;
+use jormungandr_lib::interfaces::PersistentFragmentLog;
+
+use crate::cli::recovery::tally::Error;
+use std::collections::vec_deque::VecDeque;
+
+pub fn generate_ledger_from_fragments<'block0>(
+    block0_header_id: HeaderId,
+    block0_fragments: impl Iterator<Item = &'block0 Fragment>,
+    logged_fragments: impl Iterator<Item = Result<PersistentFragmentLog, Error>>,
+) -> Result<(Ledger, VecDeque<PersistentFragmentLog>), Error> {
+    let ledger = Ledger::new(block0_header_id, block0_fragments).map_err(Error::LedgerError)?;
+    let parameters = ledger.get_ledger_parameters();
+    let block_date = ledger.date();
+
+    let mut fragments: VecDeque<PersistentFragmentLog> = VecDeque::new();
+    let mut tally_fragments: Vec<PersistentFragmentLog> = Vec::new();
+
+    // process fragments lazily
+    for fragment in logged_fragments {
+        match fragment {
+            Err(e) => {
+                println!("Error processing fragment: {:?}", e);
+            }
+            Ok(fragment_log) => {
+                if matches!(&fragment_log.fragment, Fragment::VoteTally(_)) {
+                    tally_fragments.push(fragment_log);
+                    continue;
+                }
+                match ledger.apply_fragment(&parameters, &fragment_log.fragment, block_date) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        dbg!(e);
+                        // failed to apply fragment so store for later post-process
+                        fragments.push_back(fragment_log);
+                    }
+                }
+            }
+        };
+    }
+
+    // postprocess failed fragments
+    let mut counter = 0;
+    while !fragments.is_empty() && counter < fragments.len() {
```

Comment:
I've been wondering if we should use `retain` here instead, something like... (untested, though)

```rust
    while !fragments.is_empty() {
        let len_before = fragments.len();

        fragments.retain(|fragment_log| {
            ledger
                .apply_fragment(&parameters, &fragment_log.fragment, block_date)
                .is_err()
        });

        if len_before == fragments.len() {
            break;
        }
    }
```

Reply:
Not a bad idea actually, just need to take care of logging errors here.